src/neurostatslib/solver.py

```diff
@@ -0,0 +1,593 @@
+        """
+        if not callable(func):
+            raise TypeError("The loss function must a Callable!")
+
+        if (not hasattr(func, "__module__")) or (
+            not (
+                func.__module__.startswith("jax")
+                or func.__module__.startswith("neurostatslib.glm")
+            )
+        ):
+            raise ValueError(
+                f"The function {func.__name__} is not from the jax namespace. "
+                "Only functions from the jax namespace are allowed."
+            )
+
+    @abc.abstractmethod
+    def instantiate_solver(
+        self,
+        loss: Callable[
+            [Tuple[jnp.ndarray, jnp.ndarray], jnp.ndarray, jnp.ndarray], jnp.ndarray
+        ],
+    ) -> Callable[
+        [Tuple[jnp.ndarray, jnp.ndarray], jnp.ndarray, jnp.ndarray], jaxopt.OptStep
+    ]:
+        """Abstract method to instantiate a solver with a given loss function."""
+        pass
+
+    def get_runner(
+        self,
+        solver_kwargs: dict,
+        run_kwargs: dict,
+    ) -> Callable[
+        [Tuple[jnp.ndarray, jnp.ndarray], jnp.ndarray, jnp.ndarray], jaxopt.OptStep
+    ]:
+        """
+        Get the solver runner with provided arguments.
+
+        Parameters
+        ----------
+        solver_kwargs :
+            Additional keyword arguments for the solver instantiation.
+        run_kwargs :
+            Additional keyword arguments for the solver run.
+
+        Returns
+        -------
+        :
+            The solver runner.
+        """
+        solver = getattr(jaxopt, self.solver_name)(**solver_kwargs)
+
+        def solver_run(
+            init_params: Tuple[jnp.ndarray, jnp.ndarray], X: jnp.ndarray, y: jnp.ndarray
+        ) -> jaxopt.OptStep:
+            return solver.run(init_params, X=X, y=y, **run_kwargs)
+
+        return solver_run
+
+
+class UnRegularizedSolver(Solver):
+    """
+    Solver class for optimizing unregularized models.
+
+    This class provides an interface to various optimization methods for models that
+    do not involve regularization. The optimization methods that are allowed for this
+    class are defined in the `allowed_optimizers` attribute.
+
+    Attributes
+    ----------
+    allowed_optimizers : list of str
+        List of optimizer names that are allowed for this solver class.
+
+    Methods
+    -------
+    instantiate_solver(loss)
+        Instantiates the optimization algorithm with the given loss function.
+
+    See Also
+    --------
+    [Solver](./#neurostatslib.solver.Solver) : Base solver class from which this class inherits.
+    """
+
+    allowed_optimizers = [
+        "GradientDescent",
+        "BFGS",
+        "LBFGS",
+        "ScipyMinimize",
+        "NonlinearCG",
+        "ScipyBoundedMinimize",
+        "LBFGSB",
+    ]
+
+    def __init__(
+        self, solver_name: str = "GradientDescent", solver_kwargs: Optional[dict] = None
```

Comment:
I feel like the the "no kwargs" should just an empty dict, rather than None

Reply:
the linter doesn't like mutables as default arguments

Reply:
I think mutable are annoying as arguments, because since they are pointers, and it can create unwanted behavior.

If you have None, and you reinitialize the dict all the time you avoid stuff like that to happen: let's say you want to either add something to an existing dict or you want to create a new dict if no dictionary is provided and you code it this way,

```python
def add_to_dict(key, value, my_dict={}):
    my_dict[key] = value
    return my_dict

# First call
print(add_to_dict('a', 1))  # Output: {'a': 1}

# Second call
print(add_to_dict('b', 2))  # Output: {'a': 1, 'b': 2}, not just {'b': 2}
```
It is considered better practice to have my_dict=None as default and create the dict inside the function if my_dict is None

Reply:
Oh you're totally right, I have been messed up by the mutability of default args. None it is!